### PR TITLE
Make backtraces work on Windows GNU targets again.

### PR DESCRIPTION
**Pull request by segevfiner**
This is done by adding a function that can return a filename
to pass to backtrace_create_state. The filename is obtained in
a safe way by first getting the filename, locking the file so it can't
be moved, and then getting the filename again and making sure it's the same.

See: https://github.com/rust-lang/rust/pull/37359#issuecomment-260123399
Issue: #33985

Note though that this isn't that pretty...

I had to implement a `WideCharToMultiByte` wrapper function to convert to the ANSI code page. This will work better than only allowing ASCII provided that the ANSI code page is set to the user's local language, which is often the case.

Also, please make sure that I didn't break the Unix build.

### CODE REVIEW

**Comment by petrochenkov** [comment]
@segevfiner 
Awesome!
I'll review this today.

>I had to implement a WideCharToMultiByte wrapper function to convert to the ANSI code page.

👍 
It's better to check if the string survives the `WideCharToMultiByte` -> `MultiByteToWideChar` cycle rather than work with ASCII only as https://github.com/rust-lang/rust/pull/37359#issuecomment-260123399 describes.

File: src/libstd/sys_common/gnu/libbacktrace.rs
Comment (petrochenkov) [inline src/libstd/sys_common/gnu/libbacktrace.rs]: Nit: `Ok(value)` -> `Ok((filename, file))`

File: src/libstd/sys_common/gnu/libbacktrace.rs
Comment (petrochenkov) [inline src/libstd/sys_common/gnu/libbacktrace.rs]: `static FILENAME_BUF` 1) contains a dangling reference to stack-allocated `Vec<*i8>` and 2) is not necessary

File: src/libstd/sys_common/gnu/libbacktrace.rs
Comment (petrochenkov) [inline src/libstd/sys_common/gnu/libbacktrace.rs]: Nit 2: you can do `let (filename, file) = match ::sys::backtrace::get_executable_filename() { .... }`

File: src/libstd/sys_common/gnu/libbacktrace.rs
Comment (petrochenkov) [inline src/libstd/sys_common/gnu/libbacktrace.rs]: `_file` needs to be unlocked (dropped) only if we got success from `backtrace_syminfo`.
Otherwise libbacktrace will try to read it again in next backtrace requests in unlocked state.
On failure the file needs to be either leaked or stored to be unlocked later.

File: src/libstd/sys/windows/mod.rs
Comment (petrochenkov) [inline src/libstd/sys/windows/mod.rs]: You need to validate that the reverse conversion with `MultiByteToWideChar` gives the original `u16` string, otherwise the conversion was lossy and libbacktrace will try to open some other file.

File: src/libstd/sys/windows/backtrace.rs
Comment (petrochenkov) [inline src/libstd/sys/windows/backtrace.rs]: Could you use `OpenProcess(PROCESS_QUERY_INFORMATION, FALSE, GetCurrentProcessId())` instead of `c::GetCurrentProcess()` like the https://github.com/rust-lang/rust/pull/37359 does?
`QueryFullProcessImageNameW` didn't work with pseudo-handle `-1` returned by `GetCurrentProcess` the last time I tried.

Comment (petrochenkov) [inline src/libstd/sys/windows/backtrace.rs]: It looks like there's no need to convert this `Vec<u16>` vector into `PathBuf` and back, you can just use `Vec<u16>` through all the process.

**Review by petrochenkov — CHANGES_REQUESTED** [review]

File: src/libstd/sys/windows/mod.rs
Comment (segevfiner) [inline src/libstd/sys/windows/mod.rs]: I used the `lpUsedDefaultChar` parameter of `WideCharToMultiByte`, it can save the need to convert twice assuming that it's not broken in some way.

File: src/libstd/sys/windows/backtrace.rs
Comment (segevfiner) [inline src/libstd/sys/windows/backtrace.rs]: Probably a bug they fixed in Windows 8/10 or something. Windows... 😛 

I converted it to a `PathBuf`  for the call to `OpenOptions::open` below, in a way it's a function that queries for a path so it makes sense that it returns a `Path[Buf]`.

File: src/libstd/sys_common/gnu/libbacktrace.rs
Comment (petrochenkov) [inline src/libstd/sys_common/gnu/libbacktrace.rs]:
```
let filename_ptr = filename.as_ptr();
mem::forget(filename);
(filename_ptr, Some(file))
```
is probably the simplest solution, but this is ok too.

**Review by petrochenkov — APPROVED** [review]

**Comment by petrochenkov** [comment]
Now I think I'm satisfied with this PR.
r? @alexcrichton for the final decision

File: src/libstd/sys/windows/mod.rs
Comment (retep998) [inline src/libstd/sys/windows/mod.rs]: Actually there's still ways in which data loss can occur other than default char replacement. You'll want to pass `WC_NO_BEST_FIT_CHARS`.

>For strings that require validation, such as file, resource, and user names, the application should always use the WC_NO_BEST_FIT_CHARS flag. This flag prevents the function from mapping characters to characters that appear similar but have very different semantics. In some cases, the semantic change can be extreme. For example, the symbol for "∞" (infinity) maps to 8 (eight) in some code pages.

Comment (petrochenkov) [inline src/libstd/sys/windows/mod.rs]: Ideally the set of flags should mirror whatever is used inside of (most probably) `CreateFileA` which is used inside posix `open` in CRT which is used in libbacktrace.

Comment (retep998) [inline src/libstd/sys/windows/mod.rs]: The function which is actually used to convert strings inside `A` functions is `RtlAnsiStringToUnicodeString`, not `MultiByteToWideChar`.

Comment (retep998) [inline src/libstd/sys/windows/mod.rs]: Also another point is that we're going in the *opposite* direction here, so we can't match the flags anyway.

Comment (petrochenkov) [inline src/libstd/sys/windows/mod.rs]: Ok, if we do want to "fix" this, then this will work 😄 
```
    let image_name1 = query_full_process_image_name()?;
    let ansi_image_name1 = WideCharToMultiByte(image_name1, TRY_TO_GUESS_FLAGS);
    let file = CreateFileA(ansi_image_name1)?;
    let image_name2 = query_full_process_image_name()?;
    if image_name1 != image_name2 {
        // Simultaneously make sure that the file is not moved and that encoding conversion was correct
        return Err
    }
    return ansi_image_name1;
```

Comment (segevfiner) [inline src/libstd/sys/windows/mod.rs]: [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/dd374130(v=vs.85).aspx) documents that `WC_NO_BEST_FIT_CHARS` should actually give the safety of the conversion being reversibly, that you wanted (assuming no surprises). So I added it as suggested.

File: src/libstd/sys/windows/backtrace.rs
Comment (alexcrichton) [inline src/libstd/sys/windows/backtrace.rs]: I think we should check for errors here, right?

File: src/libstd/sys/windows/backtrace.rs
Comment (alexcrichton) [inline src/libstd/sys/windows/backtrace.rs]: Out of curiosity, how much does this buy us over just attempting utf-8 and returning an error if not?

We don't use these standard functions anywhere else in the standard library, so I'm curious if we'd get much benefit from trying to do this here now.

File: src/libstd/sys_common/gnu/libbacktrace.rs
Comment (alexcrichton) [inline src/libstd/sys_common/gnu/libbacktrace.rs]: How come `file` is returned here? We're going to unconditionally reuse `STATE` so long as it's non-null, so it seems like so long as we do that we should keep the file open?

File: src/libstd/sys/windows/backtrace.rs
Comment (retep998) [inline src/libstd/sys/windows/backtrace.rs]: Because your UTF-8 bytes might be valid in some other encoding, but have different meaning, which would result in accessing the wrong file, which could potentially leave a vector open for attack.

File: src/libstd/sys/windows/backtrace.rs
Comment (segevfiner) [inline src/libstd/sys/windows/backtrace.rs]: Reading `fill_utf16_buf` carefully and seeing that `QueryFullProcessImageNameW` will leave `sz` as is on error. You are right.

File: src/libstd/sys/windows/backtrace.rs
Comment (segevfiner) [inline src/libstd/sys/windows/backtrace.rs]: Well...

UTF-8 will just break when passed to `open`/`backtrace_create_state` on any non ASCII character while the `WideCharToMultiByte(CP_ACP, ...)` will work on any character that is supported in the current ANSI code page.

The ANSI code page is a setting that you can set in Windows which is used by any non Unicode aware (wide char) program, as such it's often set to a code page for the user's local language allowing such programs to work with it. Sadly Windows doesn't use UTF-8 like Linux and keen.

As such using the ANSI code page is very likely to keep this working despite non ASCII characters in the path.

File: src/libstd/sys_common/gnu/libbacktrace.rs
Comment (segevfiner) [inline src/libstd/sys_common/gnu/libbacktrace.rs]: We only need to keep the file open/locked during the first call to a libbacktrace function that operates on the state unless that function fails and as such it will try the initialization again on the next call.

What is currently done is to optionally return the file on the first creation of the state. And if the first call to `libbacktrace_syminfo` fails we leak it so it stays open, otherwise it will be dropped on exit from `print`.

The options I can think of:
1. Refactor the code so it's clearer, I'm open to any suggestions here.
2. Leak it always. `Just mem::forget` immediately.
3. Return it and keep it in a `static mut` (Annoying since it's a drop type so we need to play around with `Box` and raw pointers), and drop it explicitly once `backtrace_syminfo` succeeds the first time.
4. The same as 3 but place it in a `static mut` immediately in `init_state`.

File: src/libstd/sys/windows/backtrace.rs
Comment (petrochenkov) [inline src/libstd/sys/windows/backtrace.rs]: >We don't use these standard functions anywhere else in the standard library

~~We~~They use them in libbacktrace :slightly_smiling_face: 
And we have to do the same.

File: src/libstd/sys_common/gnu/libbacktrace.rs
Comment (alexcrichton) [inline src/libstd/sys_common/gnu/libbacktrace.rs]: Ah ok, thanks for the explanation! I think it's probably easiest/best to just leak it here rather than special casing the result later on (depending on internal impl details of libbacktrace). We can probably use `.into_raw_handle()` here as well instead of `mem::forget`

Comment (segevfiner) [inline src/libstd/sys_common/gnu/libbacktrace.rs]: Actually this is `sys_common`. Which means I can't use `.into_raw_handle()` since it only exists on Windows (It's `.into_raw_fd()` in Unix/Linux) and `cfg!`/`#[cfg]` is not allowed in `sys_common` (Enforced by tidy). 😕

**Comment by alexcrichton** [comment]
Looks good to me, thanks @segevfiner! Want to squash the commits and I'll r+?

**Comment by segevfiner** [comment]
Squashed! Thanks for looking at this too! Now I guess it's time to cross fingers for CI. 😛

**Comment by petrochenkov** [comment]
>💔 Test failed - status-travis

`get_executable_filename` is unused on some platforms

**Comment by segevfiner** [comment]
@petrochenkov Yeah, I noticed. Working on it.

**Comment by segevfiner** [comment]
I tried to eliminate the dead code lints: (New commit so you can see what I did, instead of immediately squashing):

* Moved the get_executable_filename into sys::backgrace::gnu so I can cfg it out when it's not needed same as done for sys_common::gnu.
* I had to split the functions I added in sys/windows/backtrace.rs to sys/windows/backtrace_gnu.rs, which I hope I did correctly. (Windows has the MSVC configuration which doesn't use libbbacktrace).
* I left `multi_byte_to_wide_char` as `#[allow(dead_code]` since it would be weird to cfg it out on `target_env="gnu"`. 
* I also add to add a gnu mod to c.rs in sys/windows so that the functions used only by sys::backgrace::gnu won't trigger dead code lints.

I tested x86_64-pc-windows-gnu and x86_64-pc-windows-msvc. My computer is slow and lacking disk space to test Linux and some non-GNU Unix platform right now. Hopefully the fix there is trivial enough that I got it right this time 😞

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Comment by retep998** [comment]
Yay for c_char being signed on some platforms and unsigned on others!

**Comment by segevfiner** [comment]
Go C! Which type alias for c_char should I use here? Can I use libc::c_char?

**Comment by alexcrichton** [comment]
Yes that's what the [bindings are using](https://github.com/rust-lang/rust/blob/62a72586775fc0bfe38f4e43bf3aca52a31c6921/src/libstd/sys_common/gnu/libbacktrace.rs#L46), so that's what should be used elsewhere as inputs to here as well.

**Comment by segevfiner** [comment]
Added a commit which fixes this. I can squash to one commit if you want.

P.S. C also allows for even further evil. It requires `sizeof(char) == 1` but `CHAR_BIT` which is the number of bits in `char` doesn't have to be 8... Good luck getting anything to work on that 😝

**Comment by alexcrichton** [comment]
@bors: retry

* prioritizing rollup

**Comment by alexcrichton** [comment]
@bors: r-

I think this may have been the cause of [these failures](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1734/job/h8b3h52gg89wp40o) in the rollup

**Comment by segevfiner** [comment]
@alexcrichton It does look like backtraces are screwed up somehow in i686-pc-windows-gnu... 😖 

I will try to compile and check...

(So much platform specific fun with this pull request 😢 )

**Comment by segevfiner** [comment]
And the results are...

Build failure: #39357

**Comment by segevfiner** [comment]
I managed to build using the gcc used in the AppVeyor build and even something as simple as this:
```rust
fn f() {
    panic!("Boom!");
}

fn main() {
    f();
}
```

Doesn't display a proper stack trace in this build. The tests show weird frames with only the function drop while this minimal example seems to punt once it tries to climb one frame back.

**Comment by petrochenkov** [comment]
@segevfiner 
When backtraces worked on Windows/GNU they still were relatively buggy and could show garbage (What I see in https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1734/job/h8b3h52gg89wp40o looks very similar to that garbage).
Do `run-pass/backtrace.rs` and `run-pass/backtrace-debuginfo.rs` pass on `x86_64-pc-windows-gnu`?
If yes, then this PR is an improvement and can be merged with these tests disabled on `i686`.

**Comment by petrochenkov** [comment]
`drop::h82dbe416e9337a7a` instead of `<unknown>` means that this PR does its job.

**Comment by segevfiner** [comment]
@petrochenkov I should have looked first... https://github.com/rust-lang/rust/issues/28218 😆 

I think I managed to fix this by simply adding `base.eliminate_frame_pointer = false;` to "src\librustc_back\target\i686_pc_windows_gnu.rs".

But this completely disables this optimization. MSVC disables frame pointer optimizations by default nowadays, "/Oy-" is added by some arcane msbuild stuff. So I wonder if this is not something that is acceptable to disable so that backtraces work?

So I guess we need a choice:
1. Disable the tests and hope for someone to fix this in some way in the future. Perhaps in a different way.
2. Disable frame pointer optimizations, pay the performance price (I don't know how big it is if any) and have backtraces working in i686-pc-windows-gnu.

**Comment by petrochenkov** [comment]
@segevfiner
I think this is a separate issue that should be discussed and fixed in a separate PR and shouldn't block this PR.

**Comment by petrochenkov** [comment]
(I did some measurements with enabled/disabled frame pointers on couple of projects with MSVC+PDB toolchain, but it was years ago. There were small performance differences, but in still never disabled frame pointers in release builds because it resulted in absolutely horrible debugging experience.
Things could change. Or GNU+DWARF may behave differently. I don't know.)

**Comment by retep998** [comment]
I think we should disable frame pointer optimization on `i686-pc-windows-gnu` specifically. The problem with i686 on Windows is that, unlike all other architectures which use table based exceptions, it use stack based exception chaining, which means it kinda depends heavily on the frame pointer to be able to walk the stack.

**Comment by segevfiner** [comment]
The follow up commit which disables FPO and restores the test is at https://github.com/segevfiner/rust/commit/58c9bab24df50e8cbbfcd6122d323be754c7bab0 which I will rebase and open a PR for, once the CI will stop finding surprises, so that this can be discussed and tested properly.

**Pull request merged**
